params/json.libevm.go

```
@@ -19,98 +19,129 @@ package params
+	return MarshalChainConfigJSON(*c, c.extra, ec.reuseJSONRoot)
+}
 
-	case !reg.Get().reuseJSONRoot:
-		return c.marshalJSONWithExtra()
-
-	default: // reg.reuseJSONRoot == true
-		// The inverse of reusing the JSON root is merging two JSON buffers,
-		// which isn't supported by the native package. So we use
-		// map[string]json.RawMessage intermediates.
-		geth, err := toJSONRawMessages((*chainConfigWithoutMethods)(c))
-		if err != nil {
-			return nil, err
+// MarshalChainConfigJSON JSON encodes `config` and `extra` according to the following.
+//   - `reuseJSONRoot` is false:
+//     `config` is encoded with `extra` encoded at the "extra" JSON field.
+//   - `reuseJSONRoot` is true:
+//     `config` is encoded with `extra` encoded at the root depth of the JSON object.
+func MarshalChainConfigJSON[T any](config ChainConfig, extra T, reuseJSONRoot bool) (data []byte, err error) {
+	if !reuseJSONRoot {
+		jsonExtra := struct {
+			ChainConfig
+			Extra T `json:"extra,omitempty"`
+		}{
+			ChainConfig: config,
+			Extra:       extra,
 		}
-		extra, err := toJSONRawMessages(c.extra)
+		data, err = json.Marshal(jsonExtra)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("encoding config with extra: %s", err)
 		}
+		return data, nil
+	}
 
-		for k, v := range extra {
-			if _, ok := geth[k]; ok {
-				return nil, fmt.Errorf("duplicate JSON key %q in both %T and registered extra", k, c)
-			}
-			geth[k] = v
-		}
-		return json.Marshal(geth)
+	// The inverse of reusing the JSON root is merging two JSON buffers,
+	// which isn't supported by the native package. So we use
+	// map[string]json.RawMessage intermediates.
+	// Note we cannot encode a combined struct directly because of the extra
+	// type generic nature which cannot be embedded in such a combined struct.
+	configJSONRaw, err := toJSONRawMessages((chainConfigWithoutMethods)(config))
+	if err != nil {
+		return nil, fmt.Errorf("converting config to JSON raw messages: %s", err)
```

I think for this and the extras too we should just return a raw error and decorate it in `toJSONRawMessages()` instead, for consistency. I'd add a suggestion but GitHub review UI sucks and I can't comment on unmodified code in the same file.

`fmt.Errorf("encoding %T into %T: %s", v, msgs, err)` on line 154.

---

I think `fmt.Errorf("encoding %T into %T: %s", v, msgs, err)` is a bit confusing, since technically we are encoding then decoding into a different Go element (map). I still think `converting config to JSON raw messages` explains it alright. I've added error wrappings in 359448d66f0a15e7f62b8aaef7bd734872690975 but it might be a bit too-muchey, feel free to let me know and i'll drop that commit.